ctutil/serialization.go

```diff
@@ -181,14 +183,79 @@ func MerkleTreeLeafFromChain(chain []*x509.Certificate, etype LogEntryType, time
 		return nil, fmt.Errorf("failed to remove poison extension: %v", err)
 	}
 
-	leaf.TimestampedEntry.EntryType = PrecertLogEntryType
-	leaf.TimestampedEntry.PrecertEntry = &PreCert{
+	leaf.TimestampedEntry.EntryType = ct.PrecertLogEntryType
+	leaf.TimestampedEntry.PrecertEntry = &ct.PreCert{
 		IssuerKeyHash:  sha256.Sum256(issuer.RawSubjectPublicKeyInfo),
 		TBSCertificate: defangedTBS,
 	}
 	return &leaf, nil
 }
 
+// MerkleTreeLeafFromChainEmbeddedSCT generates a MerkleTreeLeaf from a chain and an
+// SCT timestamp, where the leaf certificate at chain[0] is a certificate that
+// has embedded within it the SCT that the timestamp is from.
+func MerkleTreeLeafFromChainEmbeddedSCT(chain []*x509.Certificate, timestamp uint64) (*ct.MerkleTreeLeaf, error) {
+	// Check that the leaf certificate does indeed contain an embedded SCT with
+	// the timestamp provided.
+	timestampOK, err := containsSCTWithTimestamp(chain[0], timestamp)
+	if err != nil {
+		return nil, err
+	}
+	if !timestampOK {
+		return nil, fmt.Errorf("leaf cert doesn't contain an embedded SCT with timestamp %d", timestamp)
+	}
+
+	leaf := ct.MerkleTreeLeaf{
+		Version:  ct.V1,
+		LeafType: ct.TimestampedEntryLeafType,
+		TimestampedEntry: &ct.TimestampedEntry{
+			EntryType: ct.PrecertLogEntryType,
+			Timestamp: timestamp,
+		},
+	}
+
+	// For building the leaf for a certificate and SCT where the SCT is embedded
+	// in the certificate, we need to build the original precertificate TBS
+	// data.  First, parse the leaf cert and its issuer.
+	if len(chain) < 2 {
+		return nil, fmt.Errorf("no issuer cert available for precert leaf building")
+	}
+	issuer := chain[1]
+	cert := chain[0]
+
+	// Next, post-process the DER-encoded TBSCertificate, to remove the SCTList
+	// extension.
+	tbs, err := x509.RemoveSCTList(cert.RawTBSCertificate)
+	if err != nil {
+		return nil, fmt.Errorf("failed to remove SCT List extension: %v", err)
+	}
+
+	leaf.TimestampedEntry.EntryType = ct.PrecertLogEntryType
+	leaf.TimestampedEntry.PrecertEntry = &ct.PreCert{
+		IssuerKeyHash:  sha256.Sum256(issuer.RawSubjectPublicKeyInfo),
+		TBSCertificate: tbs,
+	}
+	return &leaf, nil
+}
+
+func containsSCTWithTimestamp(cert *x509.Certificate, timestamp uint64) (bool, error) {
+	if cert == nil {
+		return false, nil
+	}
+
+	scts, err := x509util.ParseSCTsFromSCTList(&cert.SCTList)
```

Comment:
Is this the only line that necessitates all of the package-splitting stuff?  (I couldn't see anything else on a quick glance, but then I've not been through your import-loop pain...)

Reply:
It was this that was causing issues yes.  What I have done is removed this check from the new MerkleLeaf function, but added a comment to it saying that it assumes the timestamp is from an SCT embedded within the cert, in order to bypass the import loop issue.